File: Knull Zone Algorithm/knull_zone_algorithm.js
/**
 * Knull Zone Algorithm (Wedge Prediction Model)
 * 
 * This algorithm calculates the structural wedge risk (0 - 100) for a human extractor
 * or victim inside a tight cave constriction (like the Birth Canal or Squeeze in Nutty Putty Cave).
 * It evaluates physical dimensions relative to human body metrics (shoulder width, chest depth),
 * slope gravity vectors, and wall convergence.
 */

class KnullZoneAlgorithm {
    constructor() {
        // Average human body dimensions (in centimeters)
        this.shoulderWidth = 45.7; // Standard adult shoulder width
        this.chestDepth = 25.4;    // Standard adult chest depth
        this.hipWidth = 36.0;      // Standard adult hip width
    }

    /**
     * Predicts the wedging/stuck probability and risk components.
     * @param {Object} params 
     * @param {number} params.tunnelWidth - Width of the tunnel passage in cm
     * @param {number} params.tunnelHeight - Height of the tunnel passage in cm
     * @param {number} params.slope - Slope angle in degrees (-90 to +90)
     * @returns {Object} Risk analysis report
     */
    predict(params) {
        const { tunnelWidth, tunnelHeight, slope } = params;

        // 1. Width Clearance Ratio (horizontal clearance relative to shoulder width)
        const widthClearance = (tunnelWidth - this.shoulderWidth) / this.shoulderWidth;
        
        // 2. Height Clearance Ratio (vertical clearance relative to chest depth)
        const heightClearance = (tunnelHeight - this.chestDepth) / this.chestDepth;

        // 3. Slope Gravitational Vector Factor
        // Steeper decline/incline forces body weight into the pinch point under gravity.
        // Math: sin^1.5 of the slope angle, amplified.
        const slopeRad = Math.abs(slope) * Math.PI / 180;
        const slopeFactor = Math.pow(Math.sin(slopeRad), 1.5) * 2;

        // 4. Convergence Risk
        // If the passage width is less than the shoulder width, risk increases rapidly (compression zone).
        const convergenceRisk = widthClearance < 0 ? Math.abs(widthClearance) * 3 : 0;

        // Compute individual risk components normalized to [0, 1]
        // Width risk increases as width clearance drops below 50% extra space
        const widthRisk = Math.max(0, Math.min(1, 1 - (widthClearance + 0.3) / 0.8));
        
        // Height risk increases as height clearance drops below 40% extra space
        const heightRisk = Math.max(0, Math.min(1, 1 - (heightClearance + 0.2) / 0.6));
        
        // Slope risk capped at 1.0
        const slopeRisk = Math.min(1, slopeFactor);
        
        // Convergence risk capped at 1.0
        const convRisk = Math.min(1, convergenceRisk);

        // Weighted combination of risk factors
        // Width: 35%, Height: 20%, Slope: 25%, Convergence: 20%
        const compositeRisk = (
            widthRisk * 0.35 +
            heightRisk * 0.20 +
            slopeRisk * 0.25 +
            convRisk * 0.20
        );

        // Non-linear scaling (power curves) to make extreme tight squeezes exponentially more dangerous
        const finalScore = Math.round(
            Math.max(0, Math.min(100, Math.pow(compositeRisk, 0.8) * 100))
        );

        return {
            score: finalScore,
            components: {
                widthRisk: Math.round(widthRisk * 100),
                heightRisk: Math.round(heightRisk * 100),
                slopeRisk: Math.round(slopeRisk * 100),
                convergenceRisk: Math.round(convRisk * 100)
            },
            clearance: {
                widthClearance: (widthClearance * 100).toFixed(1),
                heightClearance: (heightClearance * 100).toFixed(1)
            },
            passable: finalScore < 60, // Verdict boundary: >= 60 is deemed a high wedge risk
            slopeAngle: Math.abs(slope)
        };
    }
}

// Export for browser compatibility (global scope) or CommonJS/ES6 environments
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { KnullZoneAlgorithm };
} else {
    window.KnullZoneAlgorithm = KnullZoneAlgorithm;
}

File: Knull Zone Algorithm/knull_zone_algorithm.py
#!/usr/bin/env python3
"""
Knull Zone Algorithm (Wedge Prediction Model)

This module implements the Knull Zone Algorithm, which calculates the structural 
wedge risk (0 - 100) for a human operator or victim stuck in narrow, sloping 
subterranean passages. 

The algorithm models the physical clearances of a human body (shoulder width, 
chest depth) relative to passage dimensions, factoring in gravity vectors from 
slope angles and wall convergence.
"""

import math

class KnullZoneAlgorithm:
    def __init__(self):
        # Average adult human dimensions (in centimeters)
        self.shoulder_width = 45.7  # Average shoulder width
        self.chest_depth = 25.4     # Average chest depth
        self.hip_width = 36.0       # Average hip width

    def predict(self, tunnel_width: float, tunnel_height: float, slope: float) -> dict:
        """
        Predicts wedging/stuck probability and returns a detailed risk analysis.
        
        :param tunnel_width: Width of the passage in centimeters
        :param tunnel_height: Height of the passage in centimeters
        :param slope: Slope angle of the passage in degrees (-90 to +90)
        :returns: Dictionary with composite score, component risks, clearances, and passability
        """
        # 1. Width clearance ratio (horizontal space relative to shoulder width)
        width_clearance = (tunnel_width - self.shoulder_width) / self.shoulder_width

        # 2. Height clearance ratio (vertical space relative to chest depth)
        height_clearance = (tunnel_height - self.chest_depth) / self.chest_depth

        # 3. Slope gravitational vector factor
        # Gravity pulls the body deeper into the wedge. Evaluated via sin^1.5 of the slope.
        slope_rad = math.radians(abs(slope))
        slope_factor = math.pow(math.sin(slope_rad), 1.5) * 2

        # 4. Convergence Risk
        # Negative clearance represents physical body compression, accelerating the wedge risk
        convergence_risk = abs(width_clearance) * 3 if width_clearance < 0 else 0.0

        # Normalize component risks to [0, 1]
        width_risk = max(0.0, min(1.0, 1.0 - (width_clearance + 0.3) / 0.8))
        height_risk = max(0.0, min(1.0, 1.0 - (height_clearance + 0.2) / 0.6))
        slope_risk = min(1.0, slope_factor)
        conv_risk = min(1.0, convergence_risk)

        # Composite risk: Width (35%), Height (20%), Slope (25%), Convergence (20%)
        composite_risk = (
            width_risk * 0.35 +
            height_risk * 0.20 +
            slope_risk * 0.25 +
            conv_risk * 0.20
        )

        # Non-linear scaling to match physical reality where narrow limits become exponentially worse
        final_score = round(max(0.0, min(100.0, math.pow(composite_risk, 0.8) * 100.0)))

        return {
            "score": final_score,
            "components": {
                "width_risk_pct": round(width_risk * 100),
                "height_risk_pct": round(height_risk * 100),
                "slope_risk_pct": round(slope_risk * 100),
                "convergence_risk_pct": round(conv_risk * 100)
            },
            "clearance": {
                "width_clearance_pct": round(width_clearance * 100, 1),
                "height_clearance_pct": round(height_clearance * 100, 1)
            },
            "passable": final_score < 60,
            "slope_angle": abs(slope)
        }

if __name__ == "__main__":
    # Test suite demonstrating algorithm behavior under different scenarios
    algo = KnullZoneAlgorithm()
    
    scenarios = [
        {"name": "Safe & Wide Entrance", "width": 100.0, "height": 80.0, "slope": 5.0},
        {"name": "Moderately Tight Decline", "width": 55.0, "height": 40.0, "slope": 25.0},
        {"name": "Nutty Putty Birth Canal (The Squeeze)", "width": 42.0, "height": 24.0, "slope": 70.0},
        {"name": "Sub-shoulder Compression Zone", "width": 38.0, "height": 22.0, "slope": 85.0}
    ]

    print("=" * 60)
    print("KNULL ZONE ALGORITHM — TEST EXECUTION")
    print("=" * 60)

    for i, s in enumerate(scenarios, 1):
        res = algo.predict(s["width"], s["height"], s["slope"])
        print(f"\nScenario {i}: {s['name']}")
        print(f"  Inputs: Width={s['width']}cm, Height={s['height']}cm, Slope={s['slope']}°")
        print(f"  Result Score: {res['score']}/100")
        print(f"  Passable: {'YES' if res['passable'] else 'NO (HIGH WEDGE RISK)'}")
        print(f"  Breakdown: Width Risk={res['components']['width_risk_pct']}%, "
              f"Height Risk={res['components']['height_risk_pct']}%, "
              f"Slope Risk={res['components']['slope_risk_pct']}%, "
              f"Convergence Risk={res['components']['convergence_risk_pct']}%")
    print("-" * 60)
